File: functions/src/gemini.ts
import {
  GoogleGenerativeAI,
  SchemaType,
  type ResponseSchema,
} from '@google/generative-ai';

import { resolveAllowedCategory } from './categories';
import {
  CURRENCIES,
  DEFAULT_CURRENCY,
  normalizeCurrency,
} from './currencies';
import {
  DEFAULT_PAYMENT_METHOD,
  normalizePaymentMethod,
  PAYMENT_METHODS,
} from './payment_methods';
import {
  FALLBACK_CATEGORY_NAME,
  resolveMerchant,
  type ParsedTransaction,
} from './schema';

// Prefer lite for SMS→JSON; fall back if unavailable or overloaded.
// 2.5-flash-lite is closed to new API keys; 2.0-flash-lite is retired.
export const GEMINI_MODELS = [
  'gemini-3.1-flash-lite',
  'gemini-2.5-flash',
  'gemini-3.5-flash',
] as const;

export const GEMINI_MODEL = GEMINI_MODELS[0];
export const MIN_PARSE_CONFIDENCE = 0.5;

function buildParsedTransactionSchema(
  allowedCategories: string[],
): ResponseSchema {
  return {
    type: SchemaType.OBJECT,
    properties: {
      amount: {
        type: SchemaType.NUMBER,
        description: 'Transaction amount as a number without commas',
      },
      currency: {
        type: SchemaType.STRING,
        format: 'enum',
        enum: [...CURRENCIES],
        description:
          'ISO currency code — must be exactly one value from the allowed list',
      },
      type: {
        type: SchemaType.STRING,
        description: 'debit or credit',
      },
      merchant: {
        type: SchemaType.STRING,
        description:
          'Primary merchant or payee. Use ATM for cash withdrawals with no named merchant — never Unknown',
      },
      merchantDetails: {
        type: SchemaType.STRING,
        description: 'Secondary merchant location or detail, or Unknown',
        nullable: true,
      },
      category: {
        type: SchemaType.STRING,
        format: 'enum',
        enum: allowedCategories,
        description:
          'Expense category — must be exactly one value from the allowed list',
      },
      paymentMethod: {
        type: SchemaType.STRING,
        format: 'enum',
        enum: [...PAYMENT_METHODS],
        description:
          'Payment rail — must be exactly one value from the allowed list',
      },
      bank: {
        type: SchemaType.STRING,
        description: 'Bank name if known, otherwise Unknown',
      },
      accountId: {
        type: SchemaType.STRING,
        description: 'Masked account or card identifier from the message',
      },
      branch: {
        type: SchemaType.STRING,
        description: 'Branch name if present, otherwise Unknown',
        nullable: true,
      },
      transactionTime: {
        type: SchemaType.STRING,
        description: 'ISO 8601 datetime with timezone offset',
      },
      transactionDate: {
        type: SchemaType.STRING,
        description: 'Date in YYYY-MM-DD format',
      },
      externalId: {
        type: SchemaType.STRING,
        description: 'TID, reference, or STAN if present',
        nullable: true,
      },
      externalIdType: {
        type: SchemaType.STRING,
        description: 'tid, ref, stan, or unknown',
      },
      parseConfidence: {
        type: SchemaType.NUMBER,
        description: 'Confidence from 0 to 1',
      },
    },
    required: [
      'amount',
      'currency',
      'type',
      'merchant',
      'category',
      'paymentMethod',
      'bank',
      'accountId',
      'transactionTime',
      'transactionDate',
      'externalIdType',
      'parseConfidence',
    ],
  };
}

/** Format a Date as Pakistan-local ISO date + datetime for the parse prompt. */
export function formatPakistanNow(now: Date): {
  currentDate: string;
  currentDateTime: string;
} {
  const parts = new Intl.DateTimeFormat('en-GB', {
    timeZone: 'Asia/Karachi',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hourCycle: 'h23',
  }).formatToParts(now);

  const get = (type: Intl.DateTimeFormatPartTypes): string =>
    parts.find((p) => p.type === type)?.value ?? '00';

  const currentDate = `${get('year')}-${get('month')}-${get('day')}`;
  const currentDateTime = `${currentDate}T${get('hour')}:${get('minute')}:${get('second')}+05:00`;
  return { currentDate, currentDateTime };
}

function buildSystemPrompt(
  allowedCategories: string[],
  currentDate: string,
  currentDateTime: string,
): string {
  const categoryList = allowedCategories.join(', ');

  return `You parse transaction text into structured JSON. Inputs may be Pakistani bank SMS/email alerts OR short natural-language descriptions written by users (e.g. "spent 200 at KFC", "salary 150000", "received 500 from Ali").

Rules:
- Amounts must be numbers without commas (e.g. 5990.00 not "5,990.00").
- currency: MUST be exactly one of: [${CURRENCIES.join(', ')}]. Usually ${DEFAULT_CURRENCY} for Pakistani bank SMS; use another code only when clearly stated.
- type must be lowercase "debit" or "credit". Infer from wording when needed (spent/paid/charged → debit; received/salary/credited → credit).
- Dates must use ISO format: transactionDate as YYYY-MM-DD, transactionTime as ISO 8601 with +05:00 for Pakistan.
- If the input is a manual/natural-language entry and does not specify a date or time, use exactly this current date/time supplied by the application: transactionDate=${currentDate}, transactionTime=${currentDateTime}. Never invent another "today" and never use placeholders.
- If a bank message includes a date/time, prefer those values from the message.
- merchant: required. For cash withdrawal / ATM with no named merchant, use "ATM" — never "Unknown". Put ATM location in merchantDetails when present.
- Use "Unknown" for missing merchantDetails, branch, bank, or accountId when not inferable.
- paymentMethod: MUST be exactly one of: [${PAYMENT_METHODS.join(', ')}].
  Map specifics to these generics (e.g. card used → debit_card, credit card → credit_card,
  Raast/IBFT/account transfer → bank_transfer, JazzCash/EasyPaisa → wallet, ATM cash → atm_withdrawal).
  If not inferable, use ${DEFAULT_PAYMENT_METHOD}.
- accountId should preserve masking from bank messages (e.g. xxx1215).
- externalIdType: use tid for TID, ref for reference numbers, stan for STAN, unknown otherwise (typical for manual entries).
- category: MUST be exactly one of these values: [${categoryList}].
  Income: salary, employer/company, investment only. Person-to-person (named individual, P2P, IBFT/Raast) → Transfer, never Income.
  Examples: PSO/Shell/Total → Fuel, McDonald's/KFC → Food & Dining, ATM → Cash Withdrawal, salary → Income, Ali sent 500 → Transfer.
  If none fit, use ${FALLBACK_CATEGORY_NAME}.
- parseConfidence: 0.0-1.0 based on how clearly fields were extracted.

Examples:

Input: PKR 5,990.00 charged at PSO RANGERS>LAH for card used, from A/C xxx1215 (DHA PHASE VIII BR LHR) on 06-Jul-2026 at 11:27 TID:387522
Output: {"amount":5990,"currency":"PKR","type":"debit","merchant":"PSO RANGERS","merchantDetails":"LAH","category":"Fuel","paymentMethod":"debit_card","bank":"Unknown","accountId":"xxx1215","branch":"DHA PHASE VIII BR LHR","transactionTime":"2026-07-06T11:27:00+05:00","transactionDate":"2026-07-06","externalId":"387522","externalIdType":"tid","parseConfidence":0.95}

Input: spent 200 at KFC
Output: {"amount":200,"currency":"PKR","type":"debit","merchant":"KFC","merchantDetails":null,"category":"Food & Dining","paymentMethod":"${DEFAULT_PAYMENT_METHOD}","bank":"Unknown","accountId":"Unknown","branch":null,"transactionTime":"${currentDateTime}","transactionDate":"${currentDate}","externalId":null,"externalIdType":"unknown","parseConfidence":0.9}

Input: PKR 10,000.00 withdrawn from ATM, A/C xxx1234 on 06-Jul-2026 at 11:27
Output: {"amount":10000,"currency":"PKR","type":"debit","merchant":"ATM","merchantDetails":null,"category":"Cash Withdrawal","paymentMethod":"atm_withdrawal","bank":"Unknown","accountId":"xxx1234","branch":null,"transactionTime":"2026-07-06T11:27:00+05:00","transactionDate":"2026-07-06","externalId":null,"externalIdType":"unknown","parseConfidence":0.95}`;
}

export type ParseResult =
  | { ok: true; parsed: ParsedTransaction; model: string }
  | { ok: false; error: string; lowConfidence?: boolean };

function normalizeParsed(
  raw: Record<string, unknown>,
  allowedCategories: string[],
): ParsedTransaction {
  const category = resolveAllowedCategory(
    String(raw.category ?? FALLBACK_CATEGORY_NAME),
    allowedCategories,
  );
  const paymentMethod = normalizePaymentMethod(
    String(raw.paymentMethod ?? DEFAULT_PAYMENT_METHOD),
  );
  return {
    amount: Number(raw.amount),
    currency: normalizeCurrency(String(raw.currency ?? DEFAULT_CURRENCY)),
    type: String(raw.type ?? 'debit').toLowerCase() as ParsedTransaction['type'],
    merchant: resolveMerchant(String(raw.merchant ?? ''), category, paymentMethod),
    merchantDetails:
      raw.merchantDetails == null || raw.merchantDetails === 'Unknown'
        ? null
        : String(raw.merchantDetails),
    category,
    paymentMethod,
    bank: String(raw.bank ?? 'Unknown'),
    accountId: String(raw.accountId ?? 'Unknown'),
    branch:
      raw.branch == null || raw.branch === 'Unknown'
        ? null
        : String(raw.branch),
    transactionTime: String(raw.transactionTime ?? ''),
    transactionDate: String(raw.transactionDate ?? ''),
    externalId:
      raw.externalId == null || raw.externalId === 'Unknown'
        ? null
        : String(raw.externalId),
    externalIdType: String(
      raw.externalIdType ?? 'unknown',
    ).toLowerCase() as ParsedTransaction['externalIdType'],
    parseConfidence: Number(raw.parseConfidence ?? 0),
  };
}

/** Quota, overload, unavailable-model, and capacity errors — try the next model. */
function isRetryableModelError(message: string): boolean {
  const lower = message.toLowerCase();
  return (
    message.includes('429') ||
    message.includes('RESOURCE_EXHAUSTED') ||
    message.includes('503') ||
    message.includes('UNAVAILABLE') ||
    message.includes('404') ||
    lower.includes('high demand') ||
    lower.includes('service unavailable') ||
    lower.includes('no longer available') ||
    lower.includes('not found')
  );
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function generateWithModel(
  apiKey: string,
  modelName: string,
  rawMessage: string,
  allowedCategories: string[],
  currentDate: string,
  currentDateTime: string,
): Promise<string> {
  const genAI = new GoogleGenerativeAI(apiKey);
  const model = genAI.getGenerativeModel({
    model: modelName,
    systemInstruction: buildSystemPrompt(
      allowedCategories,
      currentDate,
      currentDateTime,
    ),
    generationConfig: {
      responseMimeType: 'application/json',
      responseSchema: buildParsedTransactionSchema(allowedCategories),
      temperature: 0.1,
    },
  });

  const result = await model.generateContent(
    `Current date/time: ${currentDateTime}\n\nParse this transaction message:\n\n${rawMessage}`,
  );
  const text = result.response.text();

  if (!text) {
    throw new Error('Gemini returned an empty response');
  }

  return text;
}

export async function parseTransaction(
  apiKey: string,
  rawMessage: string,
  allowedCategories: string[],
  now: Date = new Date(),
): Promise<ParseResult> {
  if (allowedCategories.length === 0) {
    return { ok: false, error: 'No allowed categories configured' };
  }

  const { currentDate, currentDateTime } = formatPakistanNow(now);
  let lastError = 'Unknown Gemini parse error';
  const attempts: string[] = [];

  for (const [index, modelName] of GEMINI_MODELS.entries()) {
    try {
      const text = await generateWithModel(
        apiKey,
        modelName,
        rawMessage,
        allowedCategories,
        currentDate,
        currentDateTime,
      );
      const json = JSON.parse(text) as Record<string, unknown>;
      const parsed = normalizeParsed(json, allowedCategories);

      if (parsed.parseConfidence < MIN_PARSE_CONFIDENCE) {
        return {
          ok: false,
          error: `Low parse confidence: ${parsed.parseConfidence}`,
          lowConfidence: true,
        };
      }

      if (index > 0) {
        // Surface fallback usage — a 404/"not found" here likely means an
        // earlier model in GEMINI_MODELS is misconfigured or deprecated,
        // not just transiently overloaded, and should be investigated.
        console.warn('Gemini parse succeeded on fallback model', {
          model: modelName,
          skipped: GEMINI_MODELS.slice(0, index),
          attempts,
        });
      }

      return { ok: true, parsed, model: modelName };
    } catch (error) {
      const message =
        error instanceof Error ? error.message : 'Unknown Gemini parse error';
      lastError = `[${modelName}] ${message}`;
      attempts.push(lastError);

      if (isRetryableModelError(message)) {
        console.warn('Gemini model failed, trying next fallback', {
          model: modelName,
          message,
        });
        await sleep(1500);
        continue;
      }

      console.error('Gemini parse failed with a non-retryable error', {
        model: modelName,
        message,
        attempts,
      });
      return { ok: false, error: lastError };
    }
  }

  console.error('Gemini parse exhausted all fallback models', { attempts });
  return { ok: false, error: lastError };
}

File: backend/app/services/gemini.py
"""Gemini SMS → JSON. Port of `functions/src/gemini.ts` via the REST API."""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any
from zoneinfo import ZoneInfo

import httpx

from app.db.seeds.categories import FALLBACK_CATEGORY_NAME
from app.services.currencies import CURRENCIES, DEFAULT_CURRENCY, normalize_currency
from app.services.merchant_key import resolve_merchant
from app.services.payment_methods import (
    DEFAULT_PAYMENT_METHOD,
    PAYMENT_METHODS,
    normalize_payment_method,
)

logger = logging.getLogger(__name__)

# Temporary: gemini-3.1-flash-lite is returning 503 high demand.
# 2.5-flash is first, but its free tier is 20 requests/day. Callers must
# fall through this tuple on 429/503; do not pin GEMINI_MODELS[0].
GEMINI_MODELS = (
    "gemini-2.5-flash",
    "gemini-3.1-flash-lite",
    "gemini-3.5-flash",
)
# Daily free-tier exhaustion does not recover until the quota window resets.
_quota_exhausted: set[str] = set()
MIN_PARSE_CONFIDENCE = 0.5
_KARACHI = ZoneInfo("Asia/Karachi")
_ENDPOINT = (
    "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent"
)


@dataclass(frozen=True)
class ParsedTransaction:
    amount: float
    currency: str
    type: str
    merchant: str
    merchant_details: str | None
    category: str
    payment_method: str
    bank: str
    account_id: str
    branch: str | None
    transaction_time: str
    transaction_date: str
    external_id: str | None
    external_id_type: str
    parse_confidence: float


@dataclass(frozen=True)
class ParseOk:
    parsed: ParsedTransaction
    model: str


@dataclass(frozen=True)
class ParseFail:
    error: str
    low_confidence: bool = False


ParseResult = ParseOk | ParseFail


def format_pakistan_now(now: datetime) -> tuple[str, str]:
    if now.tzinfo is None:
        now = now.replace(tzinfo=UTC)
    local = now.astimezone(_KARACHI).replace(microsecond=0)
    current_date = local.date().isoformat()
    current_datetime = local.isoformat()
    return current_date, current_datetime


def resolve_allowed_category(raw: str, allowed_names: list[str]) -> str:
    trimmed = raw.strip()
    for name in allowed_names:
        if name == trimmed:
            return name
    lower = trimmed.lower()
    for name in allowed_names:
        if name.lower() == lower:
            return name
    for name in allowed_names:
        if name == FALLBACK_CATEGORY_NAME:
            return name
    return FALLBACK_CATEGORY_NAME


def _schema(allowed_categories: list[str]) -> dict[str, Any]:
    return {
        "type": "OBJECT",
        "properties": {
            "amount": {
                "type": "NUMBER",
                "description": "Transaction amount as a number without commas",
            },
            "currency": {
                "type": "STRING",
                "format": "enum",
                "enum": list(CURRENCIES),
            },
            "type": {"type": "STRING", "description": "debit or credit"},
            "merchant": {
                "type": "STRING",
                "description": (
                    "Primary merchant or payee. Use ATM for cash withdrawals "
                    "with no named merchant — never Unknown"
                ),
            },
            "merchantDetails": {"type": "STRING", "nullable": True},
            "category": {
                "type": "STRING",
                "format": "enum",
                "enum": allowed_categories,
            },
            "paymentMethod": {
                "type": "STRING",
                "format": "enum",
                "enum": list(PAYMENT_METHODS),
            },
            "bank": {"type": "STRING"},
            "accountId": {"type": "STRING"},
            "branch": {"type": "STRING", "nullable": True},
            "transactionTime": {"type": "STRING"},
            "transactionDate": {"type": "STRING"},
            "externalId": {"type": "STRING", "nullable": True},
            "externalIdType": {"type": "STRING"},
            "parseConfidence": {"type": "NUMBER"},
        },
        "required": [
            "amount",
            "currency",
            "type",
            "merchant",
            "category",
            "paymentMethod",
            "bank",
            "accountId",
            "transactionTime",
            "transactionDate",
            "externalIdType",
            "parseConfidence",
        ],
    }


def _system_prompt(
    allowed_categories: list[str], current_date: str, current_datetime: str
) -> str:
    category_list = ", ".join(allowed_categories)
    currencies = ", ".join(CURRENCIES)
    methods = ", ".join(PAYMENT_METHODS)
    return (
        "You parse transaction text into structured JSON. Inputs may be Pakistani "
        "bank SMS/email alerts OR short natural-language descriptions written by "
        'users (e.g. "spent 200 at KFC", "salary 150000", "received 500 from Ali").\n\n'
        "Rules:\n"
        '- Amounts must be numbers without commas (e.g. 5990.00 not "5,990.00").\n'
        f"- currency: MUST be exactly one of: [{currencies}]. Usually "
        f"{DEFAULT_CURRENCY} for Pakistani bank SMS; use another code only when "
        "clearly stated.\n"
        '- type must be lowercase "debit" or "credit". Infer from wording when '
        "needed (spent/paid/charged → debit; received/salary/credited → credit).\n"
        "- Dates must use ISO format: transactionDate as YYYY-MM-DD, "
        "transactionTime as ISO 8601 with +05:00 for Pakistan.\n"
        "- If the input is a manual/natural-language entry and does not specify a "
        "date or time, use exactly this current date/time supplied by the "
        f"application: transactionDate={current_date}, "
        f'transactionTime={current_datetime}. Never invent another "today" and '
        "never use placeholders.\n"
        "- If a bank message includes a date/time, prefer those values from the "
        "message.\n"
        "- merchant: required. For cash withdrawal / ATM with no named merchant, "
        'use "ATM" — never "Unknown". Put ATM location in merchantDetails when '
        "present.\n"
        '- Use "Unknown" for missing merchantDetails, branch, bank, or accountId '
        "when not inferable.\n"
        f"- paymentMethod: MUST be exactly one of: [{methods}].\n"
        f"  If not inferable, use {DEFAULT_PAYMENT_METHOD}.\n"
        "- accountId should preserve masking from bank messages (e.g. xxx1215).\n"
        "- externalIdType: use tid for TID, ref for reference numbers, stan for "
        "STAN, unknown otherwise (typical for manual entries).\n"
        f"- category: MUST be exactly one of these values: [{category_list}].\n"
        "  Income: salary, employer/company, investment only. Person-to-person "
        "(named individual, P2P, IBFT/Raast) → Transfer, never Income.\n"
        f"  If none fit, use {FALLBACK_CATEGORY_NAME}.\n"
        "- parseConfidence: 0.0-1.0 based on how clearly fields were extracted."
    )


def _normalize(raw: dict[str, Any], allowed_categories: list[str]) -> ParsedTransaction:
    details = raw.get("merchantDetails")
    branch = raw.get("branch")
    external_id = raw.get("externalId")
    category = resolve_allowed_category(
        str(raw.get("category") or FALLBACK_CATEGORY_NAME),
        allowed_categories,
    )
    payment_method = normalize_payment_method(
        str(raw.get("paymentMethod") or DEFAULT_PAYMENT_METHOD)
    )
    return ParsedTransaction(
        amount=float(raw.get("amount") or 0),
        currency=normalize_currency(str(raw.get("currency") or DEFAULT_CURRENCY)),
        type=str(raw.get("type") or "debit").lower(),
        merchant=resolve_merchant(
            str(raw.get("merchant") or ""),
            category=category,
            payment_method=payment_method,
        ),
        merchant_details=(
            None if details is None or details == "Unknown" else str(details)
        ),
        category=category,
        payment_method=payment_method,
        bank=str(raw.get("bank") or "Unknown"),
        account_id=str(raw.get("accountId") or "Unknown"),
        branch=None if branch is None or branch == "Unknown" else str(branch),
        transaction_time=str(raw.get("transactionTime") or ""),
        transaction_date=str(raw.get("transactionDate") or ""),
        external_id=(
            None
            if external_id is None or external_id == "Unknown"
            else str(external_id)
        ),
        external_id_type=str(raw.get("externalIdType") or "unknown").lower(),
        parse_confidence=float(raw.get("parseConfidence") or 0),
    )


def _is_retryable(message: str) -> bool:
    lower = message.lower()
    return (
        "429" in message
        or "RESOURCE_EXHAUSTED" in message
        or "503" in message
        or "UNAVAILABLE" in message
        or "404" in message
        or "high demand" in lower
        or "service unavailable" in lower
        or "no longer available" in lower
        or "not found" in lower
    )


def _extract_text(payload: dict[str, Any]) -> str:
    candidates = payload.get("candidates") or []
    if not candidates:
        raise RuntimeError("Gemini returned no candidates")
    parts = ((candidates[0].get("content") or {}).get("parts")) or []
    texts = [str(part.get("text") or "") for part in parts]
    text = "".join(texts).strip()
    if not text:
        raise RuntimeError("Gemini returned an empty response")
    return text


def _mark_quota_exhausted(model_name: str, message: str) -> None:
    lower = message.lower()
    if (
        "free_tier" in lower
        or "quota exceeded" in lower
        or "exceeded your current quota" in lower
    ):
        _quota_exhausted.add(model_name)


async def generate_content(
    api_key: str,
    body: dict[str, Any],
    *,
    request_timeout: float = 45.0,
) -> tuple[dict[str, Any], str]:
    """POST generateContent, walking GEMINI_MODELS on retryable errors.

    Returns (response JSON, model name). Raises RuntimeError when every
    candidate fails. A daily-quota 429 is remembered so later calls skip it.
    """
    last_error = "Unknown Gemini error"
    models = [name for name in GEMINI_MODELS if name not in _quota_exhausted]
    if not models:
        models = list(GEMINI_MODELS)

    async with httpx.AsyncClient(timeout=request_timeout) as client:
        for index, model_name in enumerate(models):
            try:
                url = _ENDPOINT.format(model=model_name)
                response = await client.post(
                    url,
                    headers={"x-goog-api-key": api_key},
                    json=body,
                )
                if response.status_code >= 400:
                    raise RuntimeError(f"{response.status_code} {response.text[:500]}")
                if index > 0:
                    logger.warning(
                        "gemini_fallback_ok",
                        extra={"model": model_name, "skipped": models[:index]},
                    )
                return response.json(), model_name
            except Exception as exc:
                message = str(exc)
                last_error = f"[{model_name}] {message}"
                if _is_retryable(message):
                    _mark_quota_exhausted(model_name, message)
                    # `message` is reserved on LogRecord; using it as extra
                    # raises KeyError and aborts the fallback loop.
                    logger.warning(
                        "gemini_model_retry",
                        extra={"model": model_name, "detail": message},
                    )
                    await asyncio.sleep(1.5)
                    continue
                logger.error(
                    "gemini_model_failed",
                    extra={"model": model_name, "detail": message},
                )
                raise RuntimeError(last_error) from exc

    logger.error("gemini_models_exhausted", extra={"detail": last_error})
    raise RuntimeError(last_error)


async def parse_transaction(
    api_key: str,
    raw_message: str,
    allowed_categories: list[str],
    now: datetime | None = None,
) -> ParseResult:
    if not allowed_categories:
        return ParseFail(error="No allowed categories configured")
    if not api_key:
        return ParseFail(error="Gemini is not configured (GEMINI_API_KEY)")

    current_date, current_datetime = format_pakistan_now(now or datetime.now(UTC))
    body = {
        "systemInstruction": {
            "parts": [
                {
                    "text": _system_prompt(
                        allowed_categories, current_date, current_datetime
                    )
                }
            ]
        },
        "contents": [
            {
                "role": "user",
                "parts": [
                    {
                        "text": (
                            f"Current date/time: {current_datetime}\n\n"
                            f"Parse this transaction message:\n\n{raw_message}"
                        )
                    }
                ],
            }
        ],
        "generationConfig": {
            "temperature": 0.1,
            "responseMimeType": "application/json",
            "responseSchema": _schema(allowed_categories),
        },
    }
    try:
        payload, model_name = await generate_content(api_key, body)
        text = _extract_text(payload)
        parsed = _normalize(json.loads(text), allowed_categories)
    except Exception as exc:
        return ParseFail(error=str(exc))
    if parsed.parse_confidence < MIN_PARSE_CONFIDENCE:
        return ParseFail(
            error=f"Low parse confidence: {parsed.parse_confidence}",
            low_confidence=True,
        )
    return ParseOk(parsed=parsed, model=model_name)
